gcloud-java-dns/src/main/java/com/google/gcloud/dns/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1266 @@
+        for (ResourceRecordSet addition : additions) {
+          ResourceRecordSet rrset = new ResourceRecordSet();
+          rrset.setName(addition.getName());
+          rrset.setRrdatas(ImmutableList.copyOf(addition.getRrdatas()));
+          rrset.setTtl(addition.getTtl());
+          rrset.setType(addition.getType());
+          String id = getUniqueId(copy.keySet());
+          copy.put(id, rrset);
+        }
+      }
+      // make it immutable and replace
+      boolean success = dnsRecords.compareAndSet(original, ImmutableSortedMap.copyOf(copy));
+      if (success) {
+        break; // success if no other thread modified the value in the meantime
+      }
+    }
+    change.setStatus("done");
+  }
+
+  /**
+   * Lists zones. Next page token is the last listed zone name and is returned only of there is more
+   * to list.
+   */
+  @VisibleForTesting
+  Response listZones(String projectId, String query) {
+    Map<String, Object> options = OptionParsers.parseListZonesOptions(query);
+    Response response = checkListOptions(options);
+    if (response != null) {
+      return response;
+    }
+    ConcurrentSkipListMap<String, ZoneContainer> containers = findProject(projectId).zones();
+    String[] fields = (String[]) options.get("fields");
+    String dnsName = (String) options.get("dnsName");
+    String pageToken = (String) options.get("pageToken");
+    Integer maxResults = options.get("maxResults") == null
+        ? null : Integer.valueOf((String) options.get("maxResults"));
+    boolean sizeReached = false; // maximum result size was reached, we should not return more
+    boolean hasMorePages = false; // should next page token be included in the response?
+    LinkedList<String> serializedZones = new LinkedList<>();
+    String lastZoneName = null;
+    ConcurrentNavigableMap<String, ZoneContainer> fragment =
+        pageToken != null ? containers.tailMap(pageToken, false) : containers;
+    for (ZoneContainer zoneContainer : fragment.values()) {
+      ManagedZone zone = zoneContainer.zone();
+      if (dnsName == null || zone.getDnsName().equals(dnsName)) {
+        if (sizeReached) {
+          // we do not add this, just note that there would be more and there should be a token
+          hasMorePages = true;
+          break;
+        } else {
+          try {
+            lastZoneName = zone.getName();
+            serializedZones.addLast(jsonFactory.toString(
+                OptionParsers.extractFields(zone, fields)));
+          } catch (IOException e) {
+            return Error.INTERNAL_ERROR.response(String.format(
+                "Error when serializing managed zone %s in project %s", zone.getName(),
+                projectId));
+          }
+        }
+      }
+      sizeReached = maxResults != null && maxResults.equals(serializedZones.size());
+    }
+    boolean includePageToken =
+        hasMorePages && (fields == null || ImmutableList.copyOf(fields).contains("nextPageToken"));
```

Comment:
You can avoid copying elements by using `Arrays.asList(fields)` instead of `ImmutableList.copyOf(fields)`.  Applies here and for other list methods.